tokio-threadpool/src/blocking/global.rs

```diff
@@ -126,56 +152,50 @@ pub fn blocking<F, T>(f: F) -> Poll<T, BlockingError>
 where
     F: FnOnce() -> T,
 {
-    let res = Worker::with_current(|worker| {
-        let worker = match worker {
-            Some(worker) => worker,
-            None => {
-                return Err(BlockingError { _p: () });
-            }
-        };
-
-        // Transition the worker state to blocking. This will exit the fn early
-        // with `NotReady` if the pool does not have enough capacity to enter
-        // blocking mode.
-        worker.transition_to_blocking()
-    });
+    CURRENT.with(|cell| {
+        let blocking = cell.get();
 
-    // If the transition cannot happen, exit early
-    try_ready!(res);
-
-    // Currently in blocking mode, so call the inner closure.
-    //
-    // "Exit" the current executor in case the blocking function wants
-    // to call a different executor.
-    let ret = tokio_executor::exit(move || f());
+        // Object-safety workaround: the `Blocking` trait must be object-safe,
+        // since we use a trait object in the thread-local. However, a blocking
+        // _operation_ will be generic over the return type of the blocking
+        // function. Therefore, rather than passing a function with a return
+        // type to `Blocking::run_blocking`, we pass a _new_ closure which
+        // doesn't have a return value. That closure invokes the blocking
+        // function and assigns its value to `ret`, which we then unpack when
+        // the blocking call finishes.
```

Comment:
Ah, yes. This is the reason why the 0.2 version doesn't actually move the `f` anywhere, and instead executes it directly in the `in_place` function. I wonder if we could do something similar here.. But that may be too large a change to worry about it.